File: golang/benchmark.go
package main

import (
	"fmt"
	"os"
	"runtime"
	"strings"
	"time"
)

type Benchmark interface {
	Prepare()
	Run(int)
	Name() string
	Warmup(Benchmark)
	Checksum() uint32
	Iterations() int
	WarmupIterations() int
	ExpectedChecksum() int64
}

type BaseBenchmark struct {
	className string
}

func (b *BaseBenchmark) ConfigVal(field string) int64 {
	return configI64(b.className, field)
}

func (b *BaseBenchmark) ConfigStr(field string) string {
	return configS(b.className, field)
}

func (b *BaseBenchmark) WarmupIterations() int {
	if cfg, ok := CONFIG[b.className]; ok {
		if n, ok := cfg["warmup_iterations"]; ok {
			if n.IsInt {
				return int(n.IntValue)
			}
		}
	}

	iter := b.Iterations()
	warmup := int(float64(iter) * 0.2)
	if warmup < 1 {
		warmup = 1
	}
	return warmup
}

func Warmup(bench Benchmark) {
	bench.Warmup(bench)
}

func (b *BaseBenchmark) Prepare() {
}

func (b *BaseBenchmark) Name() string {
	return b.className
}

func (b *BaseBenchmark) Warmup(bench Benchmark) {
	wi := b.WarmupIterations()
	for i := 0; i < wi; i++ {
		bench.Run(i)
	}
}

func (b *BaseBenchmark) Run(iteration_id int) {
}

func (b *BaseBenchmark) Checksum() uint32 {
	return 0
}

func (b *BaseBenchmark) Iterations() int {
	return int(b.ConfigVal("iterations"))
}

func RunAll(bench Benchmark) {
	for i := 0; i < bench.Iterations(); i++ {
		bench.Run(i)
	}
}

func (b *BaseBenchmark) ExpectedChecksum() int64 {
	return b.ConfigVal("checksum")
}

func RunBenchmarks(singleBench string) {
	fmt.Printf("start: %d\n", time.Now().UnixMilli())

	summaryTime := 0.0
	ok_count := 0
	fails := 0

	singleBench = strings.ToLower(singleBench)

	benchMap := map[string]Benchmark{
		"Binarytrees::Obj":        &BinarytreesObj{BaseBenchmark: BaseBenchmark{className: "Binarytrees::Obj"}},
		"Binarytrees::Arena":      &BinarytreesArena{BaseBenchmark: BaseBenchmark{className: "Binarytrees::Arena"}},
		"Brainfuck::Array":        &BrainfuckArray{BaseBenchmark: BaseBenchmark{className: "Brainfuck::Array"}},
		"Brainfuck::Recursion":    &BrainfuckRecursion{BaseBenchmark: BaseBenchmark{className: "Brainfuck::Recursion"}},
		"CLBG::Fannkuchredux":     &Fannkuchredux{BaseBenchmark: BaseBenchmark{className: "CLBG::Fannkuchredux"}},
		"CLBG::Mandelbrot":        &Mandelbrot{BaseBenchmark: BaseBenchmark{className: "CLBG::Mandelbrot"}},
		"Matmul::Single":          &Matmul1T{BaseMatmul{BaseBenchmark: BaseBenchmark{className: "Matmul::Single"}}},
		"Matmul::T4":              &Matmul4T{BaseMatmul{BaseBenchmark: BaseBenchmark{className: "Matmul::T4"}}},
		"Matmul::T8":              &Matmul8T{BaseMatmul{BaseBenchmark: BaseBenchmark{className: "Matmul::T8"}}},
		"Matmul::T16":             &Matmul16T{BaseMatmul{BaseBenchmark: BaseBenchmark{className: "Matmul::T16"}}},
		"CLBG::Nbody":             &Nbody{BaseBenchmark: BaseBenchmark{className: "CLBG::Nbody"}},
		"CLBG::Spectralnorm":      &Spectralnorm{BaseBenchmark: BaseBenchmark{className: "CLBG::Spectralnorm"}},
		"Base64::Encode":          &Base64Encode{BaseBenchmark: BaseBenchmark{className: "Base64::Encode"}},
		"Base64::Decode":          &Base64Decode{BaseBenchmark: BaseBenchmark{className: "Base64::Decode"}},
		"Json::Generate":          &JsonGenerate{BaseBenchmark: BaseBenchmark{className: "Json::Generate"}},
		"Json::ParseDom":          &JsonParseDom{BaseBenchmark: BaseBenchmark{className: "Json::ParseDom"}},
		"Json::ParseMapping":      &JsonParseMapping{BaseBenchmark: BaseBenchmark{className: "Json::ParseMapping"}},
		"Etc::Sieve":              &Sieve{BaseBenchmark: BaseBenchmark{className: "Etc::Sieve"}},
		"Etc::TextRaytracer":      &TextRaytracer{BaseBenchmark: BaseBenchmark{className: "Etc::TextRaytracer"}},
		"Etc::NeuralNet":          &NeuralNet{BaseBenchmark: BaseBenchmark{className: "Etc::NeuralNet"}},
		"Sort::Quick":             &SortQuick{BaseBenchmark: BaseBenchmark{className: "Sort::Quick"}},
		"Sort::Merge":             &SortMerge{BaseBenchmark: BaseBenchmark{className: "Sort::Merge"}},
		"Sort::Self":              &SortSelf{BaseBenchmark: BaseBenchmark{className: "Sort::Self"}},
		"Graph::BFS":              &GraphPathBFS{BaseBenchmark: BaseBenchmark{className: "Graph::BFS"}},
		"Graph::DFS":              &GraphPathDFS{BaseBenchmark: BaseBenchmark{className: "Graph::DFS"}},
		"Graph::AStar":            &GraphPathAStar{BaseBenchmark: BaseBenchmark{className: "Graph::AStar"}},
		"Hash::SHA256":            &BufferHashSHA256{BaseBenchmark: BaseBenchmark{className: "Hash::SHA256"}},
		"Hash::CRC32":             &BufferHashCRC32{BaseBenchmark: BaseBenchmark{className: "Hash::CRC32"}},
		"Etc::CacheSimulation":    &CacheSimulation{BaseBenchmark: BaseBenchmark{className: "Etc::CacheSimulation"}},
		"Calculator::Ast":         &CalculatorAst{BaseBenchmark: BaseBenchmark{className: "Calculator::Ast"}},
		"Calculator::Interpreter": &CalculatorInterpreter{BaseBenchmark: BaseBenchmark{className: "Calculator::Interpreter"}},
		"Etc::GameOfLife":         &GameOfLife{BaseBenchmark: BaseBenchmark{className: "Etc::GameOfLife"}},
		"Maze::Generator":         &MazeGenerator{BaseBenchmark: BaseBenchmark{className: "Maze::Generator"}},
		"Maze::BFS":               &MazeBFS{BaseBenchmark: BaseBenchmark{className: "Maze::BFS"}},
		"Maze::AStar":             &MazeAStar{BaseBenchmark: BaseBenchmark{className: "Maze::AStar"}},
		"Compress::BWTEncode":     &BWTEncode{BaseBenchmark: BaseBenchmark{className: "Compress::BWTEncode"}},
		"Compress::BWTDecode":     &BWTDecode{BaseBenchmark: BaseBenchmark{className: "Compress::BWTDecode"}},
		"Compress::HuffEncode":    &HuffEncode{BaseBenchmark: BaseBenchmark{className: "Compress::HuffEncode"}},
		"Compress::HuffDecode":    &HuffDecode{BaseBenchmark: BaseBenchmark{className: "Compress::HuffDecode"}},
		"Compress::ArithEncode":   &ArithEncode{BaseBenchmark: BaseBenchmark{className: "Compress::ArithEncode"}},
		"Compress::ArithDecode":   &ArithDecode{BaseBenchmark: BaseBenchmark{className: "Compress::ArithDecode"}},
		"Compress::LZWEncode":     &LZWEncode{BaseBenchmark: BaseBenchmark{className: "Compress::LZWEncode"}},
		"Compress::LZWDecode":     &LZWDecode{BaseBenchmark: BaseBenchmark{className: "Compress::LZWDecode"}},
		"Distance::Jaro":          &Jaro{BaseBenchmark: BaseBenchmark{className: "Distance::Jaro"}},
		"Distance::NGram":         &NGram{BaseBenchmark: BaseBenchmark{className: "Distance::NGram"}},
		"Etc::Words":              &Words{BaseBenchmark: BaseBenchmark{className: "Etc::Words"}},
		"Etc::LogParser":          &LogParser{BaseBenchmark: BaseBenchmark{className: "Etc::LogParser"}},
		"Template::Regex":         &TemplateRegex{TemplateBase: TemplateBase{BaseBenchmark: BaseBenchmark{className: "Template::Regex"}}},
		"Template::Parse":         &TemplateParse{TemplateBase: TemplateBase{BaseBenchmark: BaseBenchmark{className: "Template::Parse"}}},
		"CSV::Parse":              &CsvParse{BaseBenchmark: BaseBenchmark{className: "CSV::Parse"}},
	}

	for _, className := range ORDER {
		if singleBench != "" && !strings.Contains(strings.ToLower(className), singleBench) {
			continue
		}

		bench, ok := benchMap[className]
		if !ok {
			fmt.Printf("Warning: Benchmark '%s' defined in config but not found in code\n", className)
			continue
		}

		fmt.Printf("%s: ", className)

		Reset()
		bench.Prepare()
		Warmup(bench)
		runtime.GC()

		Reset()

		start := time.Now()
		RunAll(bench)
		elapsed := time.Since(start).Seconds()

		chks := bench.Checksum()
		expected := uint32(bench.ExpectedChecksum())

		if chks == expected {
			fmt.Printf("OK ")
			ok_count++
		} else {
			fmt.Printf("ERR[actual=%d, expected=%d] ", chks, expected)
			fails++
		}

		runtime.GC()

		fmt.Printf("in %.3fs\n", elapsed)
		summaryTime += elapsed
	}

	fmt.Printf("Summary: %.4fs, %d, %d, %d\n", summaryTime, ok_count+fails, ok_count, fails)

	os.WriteFile("/tmp/recompile_marker", []byte("RECOMPILE_MARKER_0"), 0644)
	if fails > 0 {
		os.Exit(1)
	}
}

File: cpp/src/benchmark.cpp
#include "benchmark.hpp"
#include "helper.hpp"
#include <chrono>
#include <cstdint>
#include <functional>
#include <iomanip>
#include <iostream>
#include <memory>
#include <re2/re2.h>
#include <thread>
#include <unordered_map>

#include "base64.hpp"
#include "binarytrees.hpp"
#include "brainfuck_array.hpp"
#include "brainfuck_recursion.hpp"
#include "buffer_hash.hpp"
#include "cache_simulation.hpp"
#include "calculator.hpp"
#include "compress.hpp"
#include "csv_parse.hpp"
#include "distance.hpp"
#include "fannkuchredux.hpp"
#include "game_of_life.hpp"
#include "graph_path.hpp"
#include "json_bench.hpp"
#include "log_parser.hpp"
#include "mandelbrot.hpp"
#include "matmul.hpp"
#include "maze.hpp"
#include "nbody.hpp"
#include "neural_net.hpp"
#include "sieve.hpp"
#include "sort.hpp"
#include "spectralnorm.hpp"
#include "template.hpp"
#include "text_raytracer.hpp"
#include "words.hpp"

int64_t Benchmark::warmup_iterations() {
  if (config_has(name())) {
    int64_t warmup = config_i64(name(), "warmup_iterations");
    if (warmup > 0)
      return warmup;
  }

  int64_t iters = iterations();
  return std::max<int64_t>(static_cast<int64_t>(iters * 0.2), 1LL);
}

void Benchmark::warmup() {
  int64_t prepare_iters = warmup_iterations();
  for (int64_t i = 0; i < prepare_iters; i++) {
    this->run(i);
  }
}

void Benchmark::run_all() {
  int64_t iters = iterations();
  for (int64_t i = 0; i < iters; i++) {
    this->run(i);
  }
}

void Benchmark::all(const std::string &single_bench,
                    const std::string &config_file) {
  load_config(config_file);

  double summary_time = 0.0;
  int ok = 0;
  int fails = 0;

  std::unordered_map<std::string, std::function<std::unique_ptr<Benchmark>()>>
      available_benches = {
          {"Binarytrees::Obj",
           []() { return std::make_unique<BinarytreesObj>(); }},
          {"Binarytrees::Arena",
           []() { return std::make_unique<BinarytreesArena>(); }},
          {"Brainfuck::Array",
           []() { return std::make_unique<BrainfuckArray>(); }},
          {"Brainfuck::Recursion",
           []() { return std::make_unique<BrainfuckRecursion>(); }},
          {"CLBG::Fannkuchredux",
           []() { return std::make_unique<Fannkuchredux>(); }},
          {"CLBG::Mandelbrot", []() { return std::make_unique<Mandelbrot>(); }},
          {"Matmul::Single", []() { return std::make_unique<Matmul1T>(); }},
          {"Matmul::T4", []() { return std::make_unique<Matmul4T>(); }},
          {"Matmul::T8", []() { return std::make_unique<Matmul8T>(); }},
          {"Matmul::T16", []() { return std::make_unique<Matmul16T>(); }},
          {"CLBG::Nbody", []() { return std::make_unique<Nbody>(); }},
          {"CLBG::Spectralnorm",
           []() { return std::make_unique<Spectralnorm>(); }},
          {"Base64::Encode", []() { return std::make_unique<Base64Encode>(); }},
          {"Base64::Decode", []() { return std::make_unique<Base64Decode>(); }},
          {"Json::Generate", []() { return std::make_unique<JsonGenerate>(); }},
          {"Json::ParseDom", []() { return std::make_unique<JsonParseDom>(); }},
          {"Json::ParseMapping",
           []() { return std::make_unique<JsonParseMapping>(); }},
          {"Etc::Sieve", []() { return std::make_unique<Sieve>(); }},
          {"Etc::TextRaytracer",
           []() { return std::make_unique<TextRaytracer>(); }},
          {"Etc::NeuralNet", []() { return std::make_unique<NeuralNet>(); }},
          {"Etc::Words", []() { return std::make_unique<Words>(); }},
          {"Sort::Quick", []() { return std::make_unique<SortQuick>(); }},
          {"Sort::Merge", []() { return std::make_unique<SortMerge>(); }},
          {"Sort::Self", []() { return std::make_unique<SortSelf>(); }},
          {"Graph::BFS", []() { return std::make_unique<GraphPathBFS>(); }},
          {"Graph::DFS", []() { return std::make_unique<GraphPathDFS>(); }},
          {"Graph::AStar", []() { return std::make_unique<GraphPathAStar>(); }},
          {"Hash::SHA256",
           []() { return std::make_unique<BufferHashSHA256>(); }},
          {"Hash::CRC32", []() { return std::make_unique<BufferHashCRC32>(); }},
          {"Etc::CacheSimulation",
           []() { return std::make_unique<CacheSimulation>(); }},
          {"Calculator::Ast",
           []() { return std::make_unique<CalculatorAst>(); }},
          {"Calculator::Interpreter",
           []() { return std::make_unique<CalculatorInterpreter>(); }},
          {"Etc::GameOfLife", []() { return std::make_unique<GameOfLife>(); }},
          {"Maze::Generator",
           []() { return std::make_unique<MazeGenerator>(); }},
          {"Maze::BFS", []() { return std::make_unique<MazeBFS>(); }},
          {"Maze::AStar", []() { return std::make_unique<MazeAStar>(); }},
          {"Compress::BWTEncode",
           []() { return std::make_unique<BWTEncode>(); }},
          {"Compress::BWTDecode",
           []() { return std::make_unique<BWTDecode>(); }},
          {"Compress::HuffEncode",
           []() { return std::make_unique<HuffEncode>(); }},
          {"Compress::HuffDecode",
           []() { return std::make_unique<HuffDecode>(); }},
          {"Compress::ArithEncode",
           []() { return std::make_unique<ArithEncode>(); }},
          {"Compress::ArithDecode",
           []() { return std::make_unique<ArithDecode>(); }},
          {"Compress::LZWEncode",
           []() { return std::make_unique<LZWEncode>(); }},
          {"Compress::LZWDecode",
           []() { return std::make_unique<LZWDecode>(); }},
          {"Distance::Jaro", []() { return std::make_unique<Jaro>(); }},
          {"Distance::NGram", []() { return std::make_unique<NGram>(); }},
          {"Etc::LogParser", []() { return std::make_unique<LogParser>(); }},
          {"Template::Regex",
           []() { return std::make_unique<TemplateRegex>(); }},
          {"Template::Parse",
           []() { return std::make_unique<TemplateParse>(); }},
          {"CSV::Parse", []() { return std::make_unique<CsvParse>(); }},
      };

  for (const auto &bench_name : config_keys()) {
    if (!single_bench.empty() &&
        to_lower(bench_name).find(to_lower(single_bench)) ==
            std::string::npos) {
      continue;
    }

    auto it = available_benches.find(bench_name);
    if (it != available_benches.end()) {
      std::cout << bench_name << ": ";
      std::cout.flush();

      auto bench = it->second();
      Helper::reset();
      bench->prepare();
      bench->warmup();
      Helper::reset();

      auto start = std::chrono::steady_clock::now();
      bench->run_all();
      auto end = std::chrono::steady_clock::now();

      std::chrono::duration<double> duration = end - start;

      uint32_t check = bench->checksum();
      uint32_t expect = static_cast<uint32_t>(bench->expected_checksum());
      if (check == expect) {
        std::cout << "OK ";
        ok++;
      } else {
        std::cout << "ERR[actual=" << check << ", expected=" << expect << "] ";
        fails++;
      }

      std::cout << "in " << std::fixed << std::setprecision(3)
                << duration.count() << "s" << std::endl;

      summary_time += duration.count();
      std::this_thread::sleep_for(std::chrono::milliseconds(1));
    } else {
      std::cout << "Warning: Benchmark '" << bench_name
                << "' defined in config but not found in code" << std::endl;
    }
  }

  if (ok + fails > 0) {
    std::cout << "Summary: " << std::fixed << std::setprecision(4)
              << summary_time << "s, " << (ok + fails) << ", " << ok << ", "
              << fails << std::endl;
  }

  if (fails > 0) {
    std::exit(1);
  }
}